src/v/kafka/server/handlers/handler_probe.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "kafka/protocol/types.h"
+#include "ssx/metrics.h"
+
+namespace kafka {
+
+/**
+ * Stores per-handler metrics for kafka requests.
+ * And exposes them to the internal metrics endpoint.
+ */
+class handler_probe {
+public:
+    explicit handler_probe(api_key);
+    void setup_metrics();
+
+    void sample_in_progress();
+    void request_completed() {
+        sample_in_progress();
+
+        _requests_completed++;
+        _requests_in_progress--;
+    }
+    void request_errored() { _requests_errored++; }
+    void request_started() {
+        sample_in_progress();
+
+        _requests_in_progress++;
+    }
+
+private:
+    ss::metrics::metric_groups _metrics;
+    api_key _key;
```

Comment:
These two fields are only used during `setup_metrics`, and can be passed as parameters. These probes don't have independent lifetimes, so `ss::metrics::metric_groups` can be a member of `handler_probe_manager`.

Reply:
I've applied the patch you sent me to fix this.